geopandas/clip.py

```diff
@@ -0,0 +1,229 @@
+
+    clip_obj : GeoDataFrame
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a clipped subset of shp
+        that intersects with clip_obj.
+    """
+
+    # Clip multi polygons
+    clipped = _clip_line_poly(shp.explode().reset_index(level=[1]), clip_obj)
+
+    lines = clipped[
+        (clipped.geometry.type == "MultiLineString")
+        | (clipped.geometry.type == "LineString")
+    ]
+    line_diss = lines.dissolve(by=[lines.index]).drop(columns="level_1")
+
+    polys = clipped[clipped.geometry.type == "Polygon"]
+    poly_diss = polys.dissolve(by=[polys.index]).drop(columns="level_1")
+
+    return gpd.GeoDataFrame(
+        pd.concat([poly_diss, line_diss], ignore_index=True)
+    )
+
+
+def clip_shp(shp, clip_obj):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS) and will
+    be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj,
+    data from shp will be clipped to the total boundary of
+    all polygons in clip_obj.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+          Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame
+          Polygon vector layer used to clip shp.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with shp.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from shp clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clipping points (glacier locations in the state of Colorado) with
+    a polygon (the boundary of Rocky Mountain National Park):
+
+        >>> import geopandas as gpd
+        >>> import earthpy.clip as cl
+        >>> from earthpy.io import path_to_example
+        >>> rmnp = gpd.read_file(path_to_example('rmnp.shp'))
+        >>> glaciers = gpd.read_file(path_to_example('colorado-glaciers.geojson'))
+        >>> glaciers.shape
+        (134, 2)
+        >>> rmnp_glaciers = cl.clip_shp(glaciers, rmnp)
+        >>> rmnp_glaciers.shape
+        (36, 2)
+
+    Clipping a line (the Continental Divide Trail) with a
+    polygon (the boundary of Rocky Mountain National Park):
+
+        >>> cdt = gpd.read_file(path_to_example('continental-div-trail.geojson'))
+        >>> rmnp_cdt_section = cl.clip_shp(cdt, rmnp)
+        >>> cdt['geometry'].length > rmnp_cdt_section['geometry'].length
+        0    True
+        dtype: bool
+
+    Clipping a polygon (Colorado counties) with another polygon
+    (the boundary of Rocky Mountain National Park):
+
+        >>> counties = gpd.read_file(path_to_example('colorado-counties.geojson'))
+        >>> counties.shape
+        (64, 13)
+        >>> rmnp_counties = cl.clip_shp(counties, rmnp)
+        >>> rmnp_counties.shape
+        (4, 13)
+    """
+    try:
```

Comment:
But we could check that the input is a GeoDataFrame (and eg not just DataFrame), that's something we now do in sjoin:

https://github.com/geopandas/geopandas/blob/b71ab0bf3a0a88151a827189dba5c97284460b5d/geopandas/tools/sjoin.py#L35-L43

Reply:
ok so what would you guys prefer? @jorisvandenbossche  would you like for me to check that it's an GDF rather than a ducktyping approach here ? Happy to make this change. The goal is to ensure that the input gdf object is a geodataframe! Please advise!

Reply:
Go for @jorisvandenbossche's suggestion above.

Reply:
ok this is interesting. as i'm working on this i realize that the code above could potentially fail. i'm providing it with a geoseries which will clip properly but it's not of type gpd.geopandas.GeoDataFrame. 

I think that is why i checked for geometry initially.  thoughts on how to handle this? @martinfleis @jorisvandenbossche

Reply:
`if not isinstance(df, (GeoDataFrame, GeoSeries))` should work if I am no mistaken.